### PR TITLE
Skeleton production project with packaging

### PR DESCRIPTION
**Pull request by MatthewHambley**
This should build a package which can be uploaded directly to PyPI. That is untested and for another ticket.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    for line in handle:
+        bits = line.split('=', 1)
+        if bits[0].strip().lower() == '__version__':
+            _version = bits[1].strip()
+            break
+    else:
+        raise RuntimeError('Cannot determine package version.')
+
+setuptools.setup(
+    name='model-fab',
```
Comment (stevewardle) [inline setup.py]: Now that I see it written below I think I actually prefer `sci-fab`

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    long_description=_long_description,
+    long_description_content_type='text/markdown',
+    url='https://github.com/Metomi/fab',
+    package_dir={'': 'source'},
+    packages=setuptools.find_packages(where='source'),
+    classifiers=[
+        'Development Status :: 1 - Planning',
+        'Environment :: Console',
+        'License :: OSI Approved :: BSD License',
+        'Operating System :: OS Independent',
```
Comment (stevewardle) [inline setup.py]: I've got a feeling we may struggle to keep this truly "OS Independent" as time goes on, but we can revise this later I suppose

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        'ProgrammingLanguage :: Python :: 3',
+        'Topic :: Scientific/Engineering',
+        'Topic :: Software Development :: Build Tools'
+        ],
+    python_requires='>=3.6, <4',
+    entry_points={'console_scripts': ['fab=fab.__main__:main']},
+    project_urls={
+        'Bug reports': 'https://github.com/Metomi/fab/issues',
```
Comment (stevewardle) [inline setup.py]: Lowercase `metomi`? (and on the line below too)

File: source/fab/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+'''
+Flexible build system for scientific software.
+'''
+
+__version__ = '0.1.dev0'
```
Comment (stevewardle) [inline source/fab/__init__.py]: Just a passing thought - I think we will eventually want to go with a YYYY.MM.D style of versioning (though I doubt it matters in the build up to the first release version)

File: source/fab/__main__.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+def parse_cli() -> argparse.Namespace:
+    '''
+    Parse the command line for arguments.
+    '''
+    description='Flexible build system for scientific software.'
+    parser = argparse.ArgumentParser(add_help=False,
+                                     description=description)
+    parser.add_argument('-help', '-h', '--help', action='help',
+                        help='Print this help and exit')
+    parser.add_argument('-version', action='version', version=fab.__version__)
```
Comment (stevewardle) [inline source/fab/__main__.py]: I think convention is generally to have `--` as a prefix to any flags that aren't single character?  (and on the `-verbose` option below as well)

File: README.md [hunk 6]
```diff
@@ -12,3 +12,7 @@ features which do not target scientific software development.
 ## Licence
 
 The software is made available under a 3-clause BSD licence.
+
+## Installation
+
+You may simply use `pip install model-fab`.
```
Comment (stevewardle) [inline README.md]: As per my other comment further down - maybe `sci-fab` is better?

File: source/fab/__init__.py (hunk 4 above)
Comment (MatthewHambley) [reply source/fab/__init__.py]: Not that I'm against changing it right now, but why do you think that?

File: source/fab/__main__.py (hunk 5 above)
Comment (MatthewHambley) [reply source/fab/__main__.py]: Only on GNU projects. Other realms prefer different approaches. In the particular case of "help" I want to cover all bases I can think of. It should always be easy to get help.

File: setup.py (hunk 1 above)
Comment (MatthewHambley) [reply setup.py]: Happy to oblige.

File: setup.py (hunk 2 above)
Comment (MatthewHambley) [reply setup.py]: Fair enough. I've changed it to Posix, which includes Linux despite it not being Posix compliant.

File: setup.py (hunk 3 above)
Comment (MatthewHambley) [reply setup.py]: Quite correct.

File: source/fab/__init__.py (hunk 4 above)
Comment (stevewardle) [reply source/fab/__init__.py]: From past experience I find having the major-minor style of versioning means you are forever questioning "should this be a major release?" whereas with the date based one it feels more like it gives every release equal(ly insignificant) weighting and takes that decision out of your hands

File: source/fab/__main__.py (hunk 5 above)
Comment (stevewardle) [reply source/fab/__main__.py]: Fair enough with help, I agree... I guess we need to decide if we are to follow GNU convention or not (just thinking about a lot of the other tools we use like Rose, Cylc etc. all seem to follow it)

File: source/fab/__init__.py (hunk 4 above)
Comment (MatthewHambley) [reply source/fab/__init__.py]: I tend to think the question of whether it is major (i.e. breaks backward compatibility) or not is an important one to ask on every release. However not strongly enough to argue the point. I have changed it and made a guess at when our first release will be.

File: source/fab/__main__.py (hunk 5 above)
Comment (MatthewHambley) [reply source/fab/__main__.py]: In the case of Rose and Cylc, they sort-of follow the GNU standard. The glibc documentation has this to say: https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html

As you can see, the GNU "long options" extension specifies `--<name>=<value>` and while Rose and Cylc does support that it also supports `--<name> <value>`.

Sun were proponents of the long option with single hyphen which is reflected in Java. I tend to prefer that mode but it does mean that you can no longer use single letter options. After all does `-cat` mean "cat" or does it mean "c", "a" and "t"?

But there isn't even agreement within Posix tools, look at `find` which offers options like `-name`.

Basically we just have to choose what we like and stick to it.

Comment (stevewardle) [reply source/fab/__main__.py]: Okay well I think we have one vote for each, @GiorgiaLine for the deciding vote?

File: source/fab/__init__.py (hunk 4 above)
Comment (stevewardle) [reply source/fab/__init__.py]: Agreed

File: source/fab/__main__.py (hunk 5 above)
Comment (GiorgiaLine) [reply source/fab/__main__.py]: I think I prefer the GNU convention for this, if only because I like being able to use single letter options.

Comment (MatthewHambley) [reply source/fab/__main__.py]: As you wish.
